File: solution/04xxx/4095 - 최대 정사각형/4095.cpp
#include <iostream>
#include <algorithm>
#include <vector>
using namespace std;

#define FASTIO cin.tie(NULL); cout.tie(NULL); ios::sync_with_stdio(false);

int main() {
    FASTIO

    int n, m;
    while(cin >> n >> m) {
        if(n == 0 && m == 0) break;
        int res = 0;
        vector<vector<int>> matrix(n, vector<int>(m, 0));
        for(int i = 0; i < n; i++) for(int j = 0; j < m; j++) cin >> matrix[i][j];
        for(int i = 0; i < n; i++) {
            for(int j = 0; j < m; j++) {
                if(matrix[i][j] == 0) continue;
                else if(i == 0 || j == 0) matrix[i][j] = 1;
                else if(matrix[i-1][j-1] != 0 && matrix[i-1][j] != 0 && matrix[i][j-1] != 0) matrix[i][j] = min({matrix[i-1][j-1], matrix[i-1][j], matrix[i][j-1]}) + 1;
                res = max(res, matrix[i][j]);
            }
        }
        cout << res << "\n";
    }
}

File: solution/04xxx/4095 - 최대 정사각형/4095.py
import sys
input = sys.stdin.readline

while True:
    n, m = map(int, input().split())
    if n == 0 and m == 0: break
    matrix = [list(map(int, input().split())) for _ in range(n)]
    res = 0
    for i in range(n):
        for j in range(m):
            if matrix[i][j] == 0: continue
            elif i == 0 or j == 0: matrix[i][j] = 1
            elif matrix[i-1][j-1] != 0 and matrix[i-1][j] != 0 and matrix[i][j-1] != 0: matrix[i][j] = min(matrix[i-1][j-1], matrix[i][j-1], matrix[i-1][j]) + 1
            res = max(res, matrix[i][j])
    print(res)